project-koku/koku
# COST-801: Update gcp workflow to handle initial monthly ingest.

Jira Issue: https://issues.redhat.com/browse/COST-801

### How to test:

The first part of the test is just to see if the initial monthly ingest grabs as much data as possible for the previous month and the current month. This means that we first must add the gcp source with the `make gcp-source` command. 

After the source is created it will kick off the download, which will realize that we don't have any bills for these months yet so it will change the scan range for the big query sql to be the start & end of the previous month, and the beginning of this month to today. 

```
(koku) bash-3.2$ ls
202011_e283ca40cf2d03af6e392ce50e00a0eb_2020-11-01:2020-11-30.csv	202012_e283ca40cf2d03af6e392ce50e00a0eb_2020-12-01:2020-12-09.csv
```
The range in the report names will be picked up by the processor and the summarization pieces to make sure we process and summarize the correct ranges. 

After the initial data upload, we need to make sure our downloader changes didn't affect the outcome of the line items table:
```
```(koku) bash-3.2$ cat 202011_d81c08d4bc751634b371f12e28a628ce_2020-11-01\:2020-11-30.csv | wc -l
    1821
postgres=# SELECT COUNT(id) FROM acct10001.reporting_gcpcostentrylineitem WHERE invoice_month = '202011';
 count
-------
  1820
(1 row)
(koku) bash-3.2$ cat 202012_d81c08d4bc751634b371f12e28a628ce_2020-12-01\:2020-12-09.csv | wc -l
     449
postgres=# SELECT COUNT(id) FROM acct10001.reporting_gcpcostentrylineitem WHERE invoice_month = '202012';
 count
-------
   448
(1 row)```
```
The line items should be one less than the number of rows in the csv because of the csv header is the only row that should not be inserted. 

Now we need to check that the range goes back to the default scan range of three days after the initial monthly upload. The best way to do this without nise data is to manually update the manifest id to make the logic think it needs to download a new manifest. 

```
UPDATE public.reporting_common_costusagereportmanifest SET assembly_id='1234:1234:2012' WHERE id=2;
```
Then you can hit the download endpoint again:
```
  | [2020-12-09 21:02:36,360: INFO/ForkPoolWorker-1] [masu.processor.tasks.get_report_files(1061e652-0725-4245-837f-12545326ffd9 via b814312b-ede3-449d-aa3a-ae08f4438d2b)] Processing Report:
koku_worker       |  schema_name: acct10001
koku_worker       |  provider: GCP
koku_worker       |  provider_uuid: 1cfd71aa-f583-44d7-9cc5-bc8bc1312c52
koku_worker       |  file: /testing/pvc_dir/processing/acct10001/gcp/202011_d350c2063b9948160c11b5bce9095b45_2020-12-06:2020-12-09.csv
```
And see that our scan range change from the start of the month to: `2020-12-06:2020-12-09.csv`
